sycl/include/CL/sycl/atomic_ref.hpp

```diff
@@ -0,0 +1,676 @@
+//==----- atomic_ref.hpp - SYCL 2020 atomic_ref ----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/access/access.hpp>
+#include <CL/sycl/atomic.hpp>
+#include <CL/sycl/detail/defines.hpp>
+#include <CL/sycl/memory_enums.hpp>
+#ifdef __SYCL_DEVICE_ONLY__
+#include <CL/sycl/detail/spirv.hpp>
+#include <CL/sycl/multi_ptr.hpp>
+#endif
+
+#ifndef __SYCL_DEVICE_ONLY__
+#include <atomic>
+#endif
+#include <type_traits>
+
+__SYCL_INLINE_NAMESPACE(cl) {
```

Comment:
> The SYCL 2020 specification defines the sycl::atomic_ref template class

Why did you put SYCL 2020 template class to `cl` namespace?

Reply:
@psamolysov-intel could you please fix that?

Reply:
@bader @vladimirlaz I've removed the `cl` namespace wrapper there: https://github.com/intel/llvm/pull/4956

Reply:
@bader @vladimirlaz When I tried to address the issue and move the `atomic_ref` class from `cl::sycl::` into the `sycl::` namespace, the following errors occurs:

```
sycl\sycl\unittests\helpers/CommonRedefinitions.hpp(149): error C2872: 'sycl': ambiguous symbol
sycl\sycl\include\CL/sycl/atomic_ref.hpp(25): note: could be 'sycl'
sycl\sycl\unittests\helpers/PiMock.hpp(38): note: or       'cl::sycl'
sycl\sycl\unittests\helpers/CommonRedefinitions.hpp(149): error C2039: 'unittest': is not a member of 'sycl'
sycl\sycl\include\CL/sycl/atomic_ref.hpp(25): note: see declaration of 'sycl'
sycl\sycl\unittests\helpers/CommonRedefinitions.hpp(149): error C2872: 'sycl': ambiguous symbol
sycl\sycl\include\CL/sycl/atomic_ref.hpp(25): note: could be 'sycl'
...
```

Even though the `atomic_ref` itself is compiled fine, since the class is included into the `sycl/sycl.hpp` header, a lot of code becomes broken. Since we have introduced two namespaces: `cl::sycl::` and `sycl::` itself and `cl::` is an inline namespace, the ambiguity occurs: does `sycl` mean `cl::sycl` or just `sycl`? So, we must change all the `sycl` occurs to `cl::sycl` but as I understand our goal is to eventually replace all `cl::sycl` with `sycl` in the future. 

Let's declare `atomic_ref` in the `cl::sycl::` namespace as it is, since the `cl::` namespace is inline, the users can reference to `atomic_ref` by `sycl::atomic_ref` without any problems. Also we have no classes defined directly in `sycl::`, not `cl::sycl::` namespace yet, even extensions from `sycl/ext/intel|oneapi` are defined in the `cl::sycl::` namespace.

Reply:
@psamolysov-intel FYI @romanovvlad and @denis-kabanov are working on transition to `sycl::` namespace.

Reply:
I thought we previously decided that all SYCL APIs are available in the `::cl::sycl` namespace when the application includes `<CL/sycl.hpp>`.  This is what the SYCL 2020 spec says:

> For compatibility with SYCL 1.2.1, SYCL provides another standard header file: <CL/sycl.hpp>, which can be included in place of <sycl/sycl.hpp>. In that case, all SYCL classes, constants, types and functions defined by this specification should exist within the `::cl::sycl` C++ namespace.

Reply:
> I thought we previously decided that all SYCL APIs are available in the `::cl::sycl` namespace when the application includes `<CL/sycl.hpp>`. This is what the SYCL 2020 spec says:
> 
> > For compatibility with SYCL 1.2.1, SYCL provides another standard header file: <CL/sycl.hpp>, which can be included in place of <sycl/sycl.hpp>. In that case, all SYCL classes, constants, types and functions defined by this specification should exist within the `::cl::sycl` C++ namespace.

Does it apply to extensions like `atomic_ref` before they are promoted to SYCL core part?

Reply:
This conversation is about the version of `atomic_ref` that *is* promoted to core SYCL, though, right?

In any case, it would probably be confusing to users if extensions behaved differently than core SYCL APIs in this regard.  Unless it is difficult to implement, I'd think that extensions should be available in the `cl::sycl::ext` namespace when `<CL/sycl.hpp>` is included.

Reply:
AFAIK, SYCL 2020 extensions are available only in `::sycl::` namespace. 
Some extensions added before SYCL 2020 are available in `cl::sycl::` namespace as well.

BTW, is there any plan to drop support for `cl::` namespace in the future?

Reply:
> Some extensions added before SYCL 2020 are available in cl::sycl:: namespace as well.

This seems confusing if it is not consistent.

> BTW, is there any plan to drop support for cl:: namespace in the future?

I think we should, but I'm not sure what the other Khronos members are thinking.